### PR TITLE
Copy paste

### PR DESCRIPTION
**Pull request by BehradToghi**
Copy Paste augmentation, ready for review.

### CODE REVIEW

**Review by ravi-mosaicml — COMMENTED** [review]
I think the debug folder was committed by mistake?

Going to defer to others for reviews on algorithms...

**Comment by BehradToghi** [comment]
I did not delete the `debug_out` folder (which contains some sample outputs of CopyPaste, in case you need them in the future. Please feel free to take it out of the PR late. @Landanjs

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [inline composer/algorithms/copypaste/copypaste.py:86]: is there any way to vectorize this while loop?

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [inline composer/algorithms/copypaste/copypaste.py:104]:
```suggestion
        if random.uniform(0, 1) < configs["p"]:
            for src_instance_id in src_instance_ids:
                trg_image, trg_mask = _copypaste_instance(
                    images[i], masks[i],  images[j], masks[j], src_instance_id, configs)
```

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [inline composer/algorithms/copypaste/copypaste.py:15]:
```suggestion
import torchvision.transforms as transforms
```
I think the convention is to just use `transforms` as the module name.

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [inline composer/algorithms/copypaste/copypaste.py:237]:
```suggestion
    src_instance_mask = torch.where(src_mask == src_instance_id, src_instance_id, bg_color)
```
Maybe just inline the function since it is one line?

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [inline composer/algorithms/copypaste/copypaste.py:89]:
```suggestion
        num_instances = len(torch.unique(masks[i])) - 1
```
This can be probably inlined too?

File: composer/algorithms/copypaste/copypaste.py
Comment (eracah) [reply composer/algorithms/copypaste/copypaste.py:86]: Something like this maybe? :

```python
trg_ims = images
trg_masks = masks
perm = torch.randint(0, images.size(dim=0), (images.size(dim=0),))
src_ims = images[perm]
src_masks = masks[perm]
src_instance_masks = src_masks * torch.where(src_masks == src_instance_ids[:, None, None], 1, 0)
src_instances = torch.where(src_instance_masks[:, None, :, :] == bg_color, torch.zeros(1), src_ims)
trg_ims = torch.where(src_instance_masks[:, None, :, :] == bg_color, trg_ims, torch.zeros(1))
trg_ims += src_instances
trg_masks = torch.where(src_instance_masks == bg_color,
                           trg_masks, src_instance_masks)
```

**Review by eracah — COMMENTED** [review]
Looks pretty good! Great work, @BehradToghi. Just had a few nits and a question about vectorizing

**Comment by eracah** [comment]
> I think the debug folder was committed by mistake?
> 
> Going to defer to others for reviews on algorithms...

It's a useful reference to look at while reviewing, but yes we need to remember to remove all of the images before merging

**Comment by Landanjs** [comment]
I'm not sure if Daya needs to be a reviewer, but I can't remove him? Cory, Evan and I should be enough

**Review by dskhudia — APPROVED** [review]
Just accepting it to unblock. Didn't review.

File: composer/algorithms/copypaste/README.md
Comment (A-Jacobson) [inline composer/algorithms/copypaste/README.md:38]:
```suggestion
            loss = area * loss_fn_ce(y_hat, y) + (1 - area) * loss_fn_ce(y_hat, y_perm) + loss_fn_mse(y_hat, y_perm)
```

File: composer/algorithms/copypaste/README.md
Comment (A-Jacobson) [inline composer/algorithms/copypaste/README.md:74]:
```suggestion
Our CopyPaste implementation offers a variety of tunable hyperparameters which enable the user to control the mechanics of transformations that are applied to the copied instances as well as governing the statistics of stochastic parts of this augmentation method. However, the most impactful hyperparameter seems to be the probability of applying CopyPaste to a selected pair of source & target samples. This tunable parameter is set to `p = 0.5` by default to ensure feeding both original (non-augmented) and augmented samples to the network during a training job.
```

File: composer/algorithms/copypaste/README.md
Comment (A-Jacobson) [inline composer/algorithms/copypaste/README.md:78]:
```suggestion
During the process of copying instances from the source sample to the target sample, several considerations are applied to the to-be-copied instance.
```

File: composer/algorithms/copypaste/README.md
Comment (A-Jacobson) [inline composer/algorithms/copypaste/README.md:79]:
```suggestion
If the mask of a copied instance collides with an existing mask in the target instance, the copied mask is placed over the existing mask.
```

File: composer/algorithms/copypaste/README.md
Comment (A-Jacobson) [inline composer/algorithms/copypaste/README.md:83]:
```suggestion
> ❗ CopyPaste ignores instances that are too small
```

File: composer/algorithms/copypaste/copypaste.py
Comment (A-Jacobson) [reply composer/algorithms/copypaste/copypaste.py:15]: it's common to use either. Here's are transforms imported as T in the official docs. https://pytorch.org/vision/stable/auto_examples/plot_scripted_tensor_transforms.html#sphx-glr-auto-examples-plot-scripted-tensor-transforms-py

File: composer/algorithms/copypaste/copypaste.py
Comment (BehradToghi) [reply composer/algorithms/copypaste/copypaste.py:104]: Thanks so much for the feedback @eracah ! This change will break the logic as images are iteratively passed to _copypaste_instance()

Comment (eracah) [reply composer/algorithms/copypaste/copypaste.py:104]: Oh my bad. didn't notice that

**Comment by Landanjs** [comment]
It automatically requested a review from you @dblalock 😞

**Comment by dblalock** [comment]
@Landanjs are we still trying to merge this? I keep getting notifications about how this needs review

**Comment by Landanjs** [comment]
I was addressing comments and trying to fix pyright in my own fork at branch `landan/copy-paste`, but I couldn't finish it in time... So it might be 1.5 weeks until another update. @dblalock could you approve to prevent notifications? Austin, Cory, and Evan would be the main reviewers

**Review by dblalock — COMMENTED** [review]
<Dummy review as per Landan's suggestion>

**Comment by Landanjs** [comment]
Hi all, I was unable to get an improvement from CopyPaste on ADE20k (see [wandb runs](https://wandb.ai/mosaic-ml/behrad_copypaste_final/reports/Copy-Paste-Final--VmlldzoyNjQxNTQ0/edit?firstReport&runsetFilter)). I tried a few tweaks and adjusting hyperparameters, but without success. Since semantic segmentation specific speedups have lower priority now and this was designed for instance segmentation (which we don't have), we decided to merge it in the contrib repo so we can come back to the method in the future. [Here is the mcontrib PR](https://github.com/mosaicml/mcontrib/pull/6)

**Comment by Landanjs** [comment]
Hi all, I was unable to get an improvement from CopyPaste on ADE20k (see [wandb runs](https://wandb.ai/mosaic-ml/behrad_copypaste_final/reports/Copy-Paste-Final--VmlldzoyNjQxNTQ0/edit?firstReport&runsetFilter)). I tried a few tweaks and adjusting hyperparameters, but without success. Since semantic segmentation specific speedups have lower priority now and this was designed for instance segmentation (which we don't have), we decided to merge it in the contrib repo so we can come back to the method in the future. [Here is the mcontrib PR](https://github.com/mosaicml/mcontrib/pull/6)

**Pull request closed without merging**
